### PR TITLE
update

### PR DESCRIPTION
**Pull request by LTRavenwood**
Some Updated code to look at, main points include:

1 .beginning a program where I would Like mimic the coding for a game called clash royale. not going to release this, just for the practice

2. grid.py is my experiments with trying to make a grid in a game so that there can be spaces for something to occupy on a battlefield of sorts

3. fixed a bug where healing with a potion could cause a character's hp to go above max hp

As always, code passes are appreciated

### CODE REVIEW

File: rpg.py [hunk 1]
```diff
@@ -123,18 +123,14 @@ def act(self, players):
     def move_input(self):
         move = None
         while move is None:
-            m_input = input('>')
-            move = m_input
+            move = input('>')
```
Comment (periodically-au) [inline rpg.py]: So... what happens if `move` is not `'f', 'i'` or `'s'`...

Comment (periodically-au) [reply rpg.py]: Hint: you might want to add an `else` statement

File: rpg.py [hunk 2]
```diff
@@ -143,7 +139,10 @@ def use_item(self):
             item_input = input('>')
             if item_input == 'y':
                 if self.hp < self.max_hp:
-                    self.hp += potion.hp_restore
+                    for i in range(0, potion.hp_restore + 1):
```
Comment (periodically-au) [inline rpg.py]: You want to use `min()`
Using a `for` loop to do this is very bad in terms of runtime complexity (scales with `self.max_hp`).
When the numbers get big, this method becomes insanely slow.

File: rpg.py [hunk 3]
```diff
@@ -308,10 +307,10 @@ def run(self):
         name = input('>')
     player1 = Ally(name=name, hp=5, max_hp=5, attack=3, speed=2)
     aqua = Ally(name='Aqua', hp=7, max_hp=7, attack=4, speed=3)
-    krillin = Enemy(name='Krillin', hp=5, max_hp=5, attack=2, speed=2)
-    yamcha = Enemy(name='Yamcha', hp=6, max_hp=6, attack=3, speed=1)
+    krillin = Enemy(name='Krillin', hp=7, max_hp=7, attack=2, speed=2)
+    yamcha = Enemy(name='Yamcha', hp=7, max_hp=7, attack=3, speed=1)
     anime_male = Enemy(name='Anime Male', hp=5, max_hp=5, attack=2, speed=2)
-    potion = Item(name='potion', hp_restore=5)
+    potion = Potion(name='potion', hp_restore=5)
```
Comment (periodically-au) [inline rpg.py]: What does having a `name` to the potion really give you?

**Review by periodically-au — CHANGES_REQUESTED** [review]

**Comment by LTRavenwood** [comment]
Update:
1. changed the hp restore function from for loop, to min function

need help allowing a player to input a second style of attack

File: rpg.py [hunk 4]
```diff
@@ -24,33 +24,42 @@ def __init__(self,
...
+        :param exp: value of the current exp a player has
+        :param target_exp: value of the exp needed for a player's level to increase"""
 
     def is_alive(self) -> bool:
         """Returns true if the player if they are alive"""
         return self.hp > 0
 
+    def is_burned(self) -> bool:
+        """check to see if the player has been burned"""
+        return self.is_burned is True
```
Comment (periodically-au) [inline rpg.py]: Two things: first you don't need this. You can just directly access the attribute `is_burned`.

Also, where is `is_burned` defined?

File: rpg.py [hunk 5]
```diff
@@ -24,33 +24,42 @@ def __init__(self,
...
+
     def is_ally(self, other_player) -> bool:
         """returns true if the player is an ally"""
         return other_player.team == self.team
 
     def deal_damage(self, other_player):
         """Deals damage if the other player is alive"""
-        if other_player.is_alive():
+        if other_player.is_alive:
```
Comment (periodically-au) [inline rpg.py]: You need `()` unless you made this a `property`. `is_alive` is still a method.

Why did you change this in the first place?

File: rpg.py [hunk 6]
```diff
@@ -120,47 +129,45 @@ def act(self, players):
...
         while move is None:
-            m_input = input('>')
-            move = m_input
+            move_input = input('>')
+            move = move_input
             if move == 'f':
-                pass
-            elif move == 'i':
+                break
+            if move == 'i':
```
Comment (periodically-au) [inline rpg.py]: `elif` was correct.

File: rpg.py [hunk 7]
```diff
@@ -120,47 +129,45 @@ def act(self, players):
...
                 self.use_item()
-            elif move == 's':
+            if move == 's':
                 self.get_stats()
-                self.move_input()
             else:
-                self.move_input()
-            return move
+                return self.move_input(players)
```
Comment (periodically-au) [inline rpg.py]: This works, but still beats the point of using a while loop. Think about the looping condition.

File: rpg.py [hunk 8]
```diff
@@ -120,47 +129,45 @@ def act(self, players):
...
-                if self.hp < self.max_hp:
-                    self.hp += potion.hp_restore
-                    inventory.remove(potion.name)
-                else:
-                    print('It won\'t help')
-                    self.move_input()
-            if item_input == 'n':
-                self.move_input()
+            item_use = None
+            while item_use is None:
```
Comment (periodically-au) [inline rpg.py]: What's the point of `item_use` if you never change it? You can just do `while True:`

I would also advise against doing something like that anyways. It's a bad coding pattern to using unconditional looping then a `break`. It's hard to follow and beats the purpose of having a looping condition.

File: rpg.py [hunk 9]
```diff
@@ -120,47 +129,45 @@ def act(self, players):
...
+                item_input = input('>')
+                if item_input == 'y':
+                    if self.hp < self.max_hp:
+                        self.hp = min(self.hp + potion.hp_restore, self.max_hp)
+                        inventory.remove(potion)
+                    else:
+                        print('It won\'t help')
+                        break
+                if item_input == 'n':
```
Comment (periodically-au) [inline rpg.py]: I don't like the use of break unless it is completely necessary
.

File: rpg.py [hunk 10]
```diff
@@ -120,47 +129,45 @@ def act(self, players):
...
-                    self.move_input()
-            if item_input == 'n':
-                self.move_input()
+            item_use = None
+            while item_use is None:
+                item_input = input('>')
+                if item_input == 'y':
+                    if self.hp < self.max_hp:
+                        self.hp = min(self.hp + potion.hp_restore, self.max_hp)
+                        inventory.remove(potion)
```
Comment (periodically-au) [inline rpg.py]: Note that you will loop again if you go down this logic.

File: rpg.py [hunk 11]
```diff
@@ -246,18 +254,22 @@ def level_up(self):
...
                     player.level += 1
                     player.max_hp += 2
                     player.hp = player.max_hp
                     player.attack += 1
                     player.speed += 1
-                    player.target_exp *= 2
+                    if player.leveling_rate == 1:
+                        player.level_exp *= 2
+                    elif player.leveling_rate == 2:
+                        player.level_exp *= 4
```
Comment (periodically-au) [inline rpg.py]: At this point you should use a formula: `player.level_exp *= (player.leveling_rate * 2)`

**Review by periodically-au — COMMENTED** [review]
Just some comments. 

`while` loops are strong tools, but you want to use them carefully. They are great for asserting certain conditions are `False` when you exit the loop (since by definition you can't). You need to consider the following when using one:
1. What must be `False` for the code after the loop to work?
2. Is there a logical execution path inside of the loop that can make the required condition (1.) `False`?
3. Does the code inside of the loop do exactly what is needed to make (1.) `False`? It is bad to be doing more with the loop than enforcing a condition.

Think about it 👍

File: Clash_Royale_Copy.py [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+        self.name = name
+        self.hp = hp
+        self.damage = damage
+        self.team = team
+
+    def is_alive(self) -> bool:
+        """returns true if the crown tower is not destroyed"""
+        return self.hp > 0
+
+    def is_ally(self, other_object):
```
Comment (periodically-au) [inline Clash_Royale_Copy.py]: Type Annotate.

Also there's a format that Pycharm provides naturally for parameter definitions and return values in docstrings

File: Clash_Royale_Copy.py [hunk 13]
```diff
@@ -0,0 +1,27 @@
+
```
Comment (periodically-au) [inline Clash_Royale_Copy.py:2]: File names should not be capitalized.

File: Clash_Royale_Copy.py [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+    def is_alive(self) -> bool:
+        """returns true if the crown tower is not destroyed"""
+        return self.hp > 0
+
+    def is_ally(self, other_object):
+        """returns true if another object, eg. the opposite crown tower or a card is yours"""
+        return other_object.team == self.team
+
+    def deal_damage(self, other_object):
+        """deals Damage to the other player's cards"""
```
Comment (periodically-au) [inline Clash_Royale_Copy.py]: Type Annotate

File: rpg.py [hunk 15]
```diff
@@ -50,34 +57,27 @@ def deal_damage(self, other_player):
         """Deals damage if the other player is alive"""
         if other_player.is_alive():
             other_player.hp -= self.attack
-            print(f'{self.name} attacked {other_player.name} for {self.attack} damage')
+            print(f'{self.name} attacked {other_player.name} for {self.attack} damage!')
+        return other_player
+
+    def deal_damage2(self, other_player):
```
Comment (periodically-au) [inline rpg.py]: Type Annotate.

Comment (periodically-au) [reply rpg.py]: Also this is not a good pattern, to reimplement an attack method. At this point you need to find an abstraction of this method, create an implementation at handles all variations you are/might use, and then create other implementation that do that.

I'll break it down a little for you:
```
def deal_damage(self, other_player: 'Character', damage_multipler: float=1.0, burn_chance: float=0.0) -> 'Character':
        """Base attack method for dealing damage to another character

        :param other_player: target of damage
        :param damage_multipler: Multipler for damage dealt
        :param burn_chance: Probability of causing a burn state on opponent
        :return: damaged player object"""
        if other_player.is_alive():
            # Apply the multipler and deal damage
            other_player.hp -= (self.attack * damage_multipler)
            # Update the player's burn state, using the OR operation to make sure 
            # That a player that is already burned doesn't have their status overriden
            other_player.is_burned = other_player.is_burned or (burn_chance > random.random()) 
        return other_player
```

Comment (periodically-au) [reply rpg.py]: Excusing the formatting, but do you see have you can build your new attack variations by just varying the arguments on a super method? This is basically sub-classing for methods.

File: rpg.py [hunk 16]
```diff
@@ -50,34 +57,27 @@ def deal_damage(self, other_player):
...
+        """Deals a bit more damage and has a chance to burn the target"""
+        if other_player.is_alive():
+            self.attack *= 2
```
Comment (periodically-au) [inline rpg.py]: This will permanently double the attacking player's attack. Do you want this?

File: rpg.py [hunk 17]
```diff
@@ -50,34 +57,27 @@ def deal_damage(self, other_player):
...
+            other_player.hp -= self.attack
+            burn_chance = random.randint(1, 100)
```
Comment (periodically-au) [inline rpg.py]: Not a bug, but probabilties should be floats in `[0,1)` range. Use `random.random()`

File: rpg.py (hunk 16 above)
Comment (periodically-au) [reply rpg.py]: Also if you use a hardcoded value twice, just make a constant variable at the top of the file and use that.

File: rpg.py [hunk 18]
```diff
@@ -50,34 +57,27 @@ def deal_damage(self, other_player):
...
+            if burn_chance < 11:
```
Comment (periodically-au) [inline rpg.py]: This is a weird arbitrary constant. Maybe make this a constant at the top of the file or a parameter of the character.

File: rpg.py [hunk 19]
```diff
@@ -50,34 +57,27 @@ def deal_damage(self, other_player):
...
+                other_player.is_burned = True
+                print(f'{self.name} attacked {other_player.name} for {self.attack} damage!')
```
Comment (periodically-au) [inline rpg.py]: This `print` will trigger only if you burn the opponent.

File: rpg.py [hunk 20]
```diff
@@ -50,34 +57,27 @@ def deal_damage(self, other_player):
...
         return other_player
 
     def get_all_enemies(self, players: List) -> List[int]:
```
Comment (periodically-au) [inline rpg.py]: `players: List['Character']`

File: rpg.py [hunk 21]
```diff
@@ -87,80 +87,105 @@ def __init__(self,
...
-                         team=1,
-                         level=1,
+                         leveling_rate,
                          exp=0,
-                         target_exp=200)
+                         level_exp=200,
+                         team=1,
+                         level=1)
 
     def act(self, players):
```
Comment (periodically-au) [inline rpg.py]: Type annotate

File: rpg.py [hunk 22]
```diff
@@ -87,80 +87,105 @@ def __init__(self,
...
                    character for character in players if character.team != 1}
-        if all_enemy_locations:
-            if self.is_alive:
-                if all_enemy_locations.count(1):
+        if self.is_alive:
+            if all_enemy_locations:
+                print(all_enemy_locations)
+                print(enemies)
+                if len(all_enemy_locations) == 1:  # if there is only one enemy in the battle:
+                    print('What move will you use 1?')
```
Comment (periodically-au) [inline rpg.py]: `1?`?

File: rpg.py [hunk 23]
```diff
@@ -87,80 +87,105 @@ def __init__(self,
...
-                    damaged_player = self.deal_damage(targeted_player)
-                    players[targeted_index] = damaged_player
-                else:
+                    damage_input = input('>')
+                    if damage_input == '1':
+                        self.deal_damage(targeted_player)
+                        print(f'{self.name} attacked {targeted_player.name} for {self.attack} damage!')
+                    elif damage_input == '2':
+                        self.deal_damage2(targeted_player)
+                        print(f'{self.name} attacked {targeted_player.name} for {self.attack} damage!')
```
Comment (periodically-au) [inline rpg.py]: I'm guessing your issue is if the input is neither '1' nor '2'. How did we make sure an `input('>')` pulled the right value in the past? You've used similar code before.

Comment (periodically-au) [reply rpg.py]: Also, when you are getting comments (externally or internally) like "You've used similar code before", then its time to make a general method.

In your case, you need to get input from a player that must be one of a set of strings, and must repeated query until it gets one such input.

Your basic function definition would look like this:
```
def blocking_input(acceptable_responses: Set[str]) -> str:
    """Calls input('>') until user gives an input that is in acceptable_responses

    :param acceptable_responses: Set of all acceptable responses from user
    :return: output such that output is in acceptable_responses"""
```

File: rpg.py [hunk 24]
```diff
@@ -87,80 +87,105 @@ def __init__(self,
...
+
+                else:  # if there is more than one enemy in the battle:
```
Comment (periodically-au) [inline rpg.py]: There's a lot of issues with this code. You never validate `damage_input` or `user_input`, so it will go wrong if a user messed up with inputs.

File: rpg.py [hunk 25]
```diff
@@ -87,80 +87,105 @@ def __init__(self,
...
-            return move
+                print('Please input a move.')
+                move = None
 
     def use_item(self):
         if potion.name in inventory:
             print(f'would you like to use {potion.name}?')
             print('[y]es or [n] no')
             item_input = input('>')
+            print(item_input)
```
Comment (periodically-au) [inline rpg.py]: I see you left in your debugging prints. (LEARN PYCHARMS DEBUG MODE)

File: rpg.py [hunk 26]
```diff
@@ -87,80 +87,105 @@ def __init__(self,
...
             elif move == 's':
                 self.get_stats()
-                self.move_input()
             else:
-                self.move_input()
-            return move
+                print('Please input a move.')
+                move = None
 
     def use_item(self):
```
Comment (periodically-au) [inline rpg.py]: Annotate and document

File: rpg.py [hunk 27]
```diff
@@ -87,80 +87,105 @@ def __init__(self,
...
+                        if damage_input == '1':
+                            self.deal_damage(targeted_enemy)
+                            print(f'{self.name} attacked {targeted_enemy.name} for {self.attack} damage!')
+                        elif damage_input == '2':
+                            self.deal_damage2(targeted_enemy)
+                            print(f'{self.name} attacked {targeted_enemy.name} for {self.attack} damage!')
 
         return players
 
     def move_input(self):
```
Comment (periodically-au) [inline rpg.py]: Annotate and document

File: rpg.py [hunk 28]
```diff
@@ -174,27 +199,37 @@ def __init__(self,
...
-                         team=2,
-                         level=1,
+                         leveling_rate,
                          exp=0,
-                         target_exp=200)
+                         level_exp=200,
+                         team=2,
+                         level=1)
 
     def act(self, players):
```
Comment (periodically-au) [inline rpg.py]: Documentation...

**Review by periodically-au — COMMENTED** [review]
"90% documented"

All jokes aside I added some comments. I want to point the power of abstraction (namely how we use it for sub-classing). 

Also, if you find yourself writing the same code a lot, you need to make a function/method to handle that code chunk so you never need to write it again.

File: rpg.py (hunk 22 above)
Comment (LTRavenwood) [reply rpg.py]: Yeah if len is equal to 1 that means there's only one enemy in the fight

**Comment by LTRavenwood** [comment]
Traceback (most recent call last):
  File "C:/Users/risha/Desktop/game/gameforclass/gameforclass/rpg.py", line 31, in <module>
    class Character:
  File "C:/Users/risha/Desktop/game/gameforclass/gameforclass/rpg.py", line 78, in Character
    def get_all_enemies(self, players: List[Character]) -> List[int]:
NameError: name 'Character' is not defined

**Comment by LTRavenwood** [comment]
Whoops, wrong error message

Traceback (most recent call last):
  File "C:/Users/risha/Desktop/game/gameforclass/gameforclass/rpg.py", line 355, in <module>
    battle1.run()
  File "C:/Users/risha/Desktop/game/gameforclass/gameforclass/rpg.py", line 320, in run
    acting_player.act(self.players)
  File "C:/Users/risha/Desktop/game/gameforclass/gameforclass/rpg.py", line 231, in act
    self.deal_damage(other_player=targeted_player, multiplier=1.3, burn_chance=10)
  File "C:/Users/risha/Desktop/game/gameforclass/gameforclass/rpg.py", line 73, in deal_damage
    if other_player.is_alive():
AttributeError: 'int' object has no attribute 'is_alive'
